Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs

```diff
@@ -0,0 +1,1265 @@
+        /// </returns>
+        /// <param name="value">The value to be formatted. </param><param name="rowIndex">The index of the cell's parent row. </param><param name="cellStyle">The <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/> in effect for the cell.</param><param name="valueTypeConverter">A <see cref="T:System.ComponentModel.TypeConverter"/> associated with the value type that provides custom conversion to the formatted value type, or null if no such custom conversion is needed.</param><param name="formattedValueTypeConverter">A <see cref="T:System.ComponentModel.TypeConverter"/> associated with the formatted value type that provides custom conversion from the value type, or null if no such custom conversion is needed.</param><param name="context">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewDataErrorContexts"/> values describing the context in which the formatted value is needed.</param><exception cref="T:System.Exception">Formatting failed and either there is no handler for the <see cref="E:System.Windows.Forms.DataGridView.DataError"/> event of the <see cref="T:System.Windows.Forms.DataGridView"/> control or the handler set the <see cref="P:System.Windows.Forms.DataGridViewDataErrorEventArgs.ThrowException"/> property to true. The exception object can typically be cast to type <see cref="T:System.FormatException"/>.</exception>
+        // protected virtual object GetFormattedValue(object value, int rowIndex, ref DataGridViewCellStyle cellStyle, TypeConverter valueTypeConverter, TypeConverter formattedValueTypeConverter, DataGridViewDataErrorContexts context)
+
+        /// <summary>
+        /// Gets the inherited shortcut menu for the current cell.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A <see cref="T:System.Windows.Forms.ContextMenuStrip"/> if the parent <see cref="T:System.Windows.Forms.DataGridView"/>, <see cref="T:System.Windows.Forms.DataGridViewRow"/>, or <see cref="T:System.Windows.Forms.DataGridViewColumn"/> has a <see cref="T:System.Windows.Forms.ContextMenuStrip"/> assigned; otherwise, null.
+        /// </returns>
+        /// <param name="rowIndex">The row index of the current cell.</param><exception cref="T:System.ArgumentOutOfRangeException">The value of the <see cref="P:System.Windows.Forms.DataGridViewElement.DataGridView"/> property of the cell is not null and the specified <paramref name="rowIndex"/> is less than 0 or greater than the number of rows in the control minus 1. </exception><exception cref="T:System.InvalidOperationException"><see cref="P:System.Windows.Forms.DataGridViewCell.ColumnIndex"/> is less than 0, indicating that the cell is a row header cell.</exception>
+        public virtual ContextMenuStrip GetInheritedContextMenuStrip(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
```

Comment:
Again, this is incorrect precondition. See http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,af0d302fa10e7897

Basically, the precondition is: 'if DataGridView is not null then rowIndex should be from [0, Rows.Count).

And if DataGridView is null those invariants should not be true.

We can express this using following construct:

`Contract.Requires(DataGridView != null || (rowIndex >= 0 && rowIndex < DataGridView.Rows.Count);`
`Contract.Requires(DataGridView != null || (ColumnIndex >= 0);`